server/datastore/mysql/migrations/tables/20240813103800_SupportMultipleABMTokens.go

```diff
@@ -0,0 +1,192 @@
+
+	CONSTRAINT fk_abm_tokens_macos_default_team_id
+		FOREIGN KEY (macos_default_team_id) REFERENCES teams (id) ON DELETE SET NULL,
+	CONSTRAINT fk_abm_tokens_ios_default_team_id
+		FOREIGN KEY (ios_default_team_id) REFERENCES teams (id) ON DELETE SET NULL,
+	CONSTRAINT fk_abm_tokens_ipados_default_team_id
+		FOREIGN KEY (ipados_default_team_id) REFERENCES teams (id) ON DELETE SET NULL
+)`)
+	if err != nil {
+		return fmt.Errorf("failed to create table abm_tokens: %w", err)
+	}
+
+	_, err = tx.Exec(`
+ALTER TABLE host_dep_assignments
+	ADD COLUMN abm_token_id int(10) UNSIGNED NULL,
+	ADD CONSTRAINT fk_host_dep_assignments_abm_token_id
+		FOREIGN KEY (abm_token_id) REFERENCES abm_tokens(id) ON DELETE SET NULL
+`)
+	if err != nil {
+		return fmt.Errorf("failed to alter table host_dep_assignments: %w", err)
+	}
+
+	// migrate the existing ABM token (if any) to the new abm_tokens table
+	const getABM = `
+SELECT
+	value
+FROM
+	mdm_config_assets
+WHERE
+	name = ? AND deletion_uuid = ''
+`
+	txx := sqlx.Tx{Tx: tx, Mapper: reflectx.NewMapperFunc("db", sqlx.NameMapper)}
+	var token []byte
+	if err := txx.Get(&token, getABM, fleet.MDMAssetABMTokenDeprecated); err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			// nothing to migrate, exit early
+			return nil
+		}
+		return fmt.Errorf("selecting team configs: %w", err)
+	}
+
+	// get the current ABM configuration from the app config
+	const getABMCfg = `
+SELECT
+	json_value->"$.mdm"
+FROM
+	app_config_json
+LIMIT 1
+`
+	var raw json.RawMessage
+	if err := txx.Get(&raw, getABMCfg); err != nil {
+		return fmt.Errorf("select MDM config from app_config_json: %w", err)
+	}
+
+	var (
+		abmTermsExpired bool
+		abmDefaultTeam  string
+	)
+	// decode if we did get an object
+	if len(raw) > 0 && raw[0] == '{' {
+		var config map[string]interface{}
+		if err := json.Unmarshal(raw, &config); err != nil {
+			return fmt.Errorf("unmarshal appconfig: %w", err)
+		}
+
+		if s, ok := config["apple_bm_default_team"].(string); ok {
+			abmDefaultTeam = s
+		}
+		if b, ok := config["apple_bm_terms_expired"].(bool); ok {
+			abmTermsExpired = b
+		}
+	}
+
+	var defaultTeamID *uint
+	if abmDefaultTeam != "" {
+		// get the default team id
+		var id uint
+		if err := txx.Get(&id, "SELECT id FROM teams WHERE name = ?", abmDefaultTeam); err != nil {
+			if !errors.Is(err, sql.ErrNoRows) {
+				return fmt.Errorf("select default ABM team id: %w", err)
+			}
+		}
+		// only use it if the team exists
+		if id > 0 {
+			defaultTeamID = &id
+		}
+	}
+
+	// NOTE: we don't know the organization_name, apple_id and renew_at of the
+	// existing token - the only way to know is to make an Apple API call and to
+	// decrypt the token, two operations that are probably not safe to do in a DB
+	// migration. Instead, insert with empty values and add a job at Fleet
+	// startup to ensure the token's information is filled.
+	// TODO(mna): e.g. insert a one-off worker job message?
```

Comment:
How do we feel about this? This is the main remaining point for ABM tokens. My rationale for not doing this in the DB migration is as follows:

* AFAIK usually Fleet is not running when migrating the DB, so we have to be very careful with the code we use from here - only the DB is guaranteed to be up, basically.
* A failure in a DB migration usually leaves things in a bad state that often require manual intervention. If the Apple API was to fail during the migration, we'd either have to fail and leave in a possibly bad state, or continue with missing values anyway and having to fill them up later.
* It's dangerous for DB migrations to rely on structs and code in other fleet packages because this code may evolve and the DB migration must be frozen in time - it is always meant to migrate a DB stuck in the version of that migration minus one, but it will be built with whatever code is current (I feel that point is hard to state clearly, but you know what I mean?).

What I have in mind is to enqueue a new `worker` job here, and implement that relatively straightforward job that would use the token to make the [Apple API AccountDetail](https://developer.apple.com/documentation/devicemanagement/get_account_detail) request and update the `abm_tokens` table's row with the "admin_id" (apple_id), organization name, and the renew date from the token's decrypted payload. Basically the equivalent of what we do in the [`GET /abm` endpoint](https://github.com/fleetdm/fleet/blob/f78761dbf6480de3a89c1765d027ee3770e4fb78/server/service/handler.go#L739).

Reply:
sounds good to me!

Reply:
I'll add this in a follow-up PR. Had a shower thought this morning - instead of a new `worker` job, I think it could be easier to just add a check in the cron job before looping over the ABM tokens to sync devices. This job (`apple_mdm_dep_profile_assigner`) runs every 1m by default so it would be pretty much as fast as a worker job. Only thing is that this is user-configurable, but at Fleet startup (which would happen after a migration) it runs after 10s, and I have ideas on how to handle edge cases that I'll bring forth on that follow-up PR.